crates/uniplate/src/biplate.rs

```diff
@@ -0,0 +1,141 @@
+    ///
+    /// If from == to then this function should return the root as the single child.
+    fn biplate(&self) -> (Tree<To>, Box<dyn Fn(Tree<To>) -> Self>);
+
+    /// Like descend but with more general types.
+    ///
+    /// If from == to then this function does not descend. Therefore, when writing definitions it
+    /// is highly unlikely that this function should be used in the recursive case. A common
+    /// pattern is to first match the types using descendBi, then continue the recursion with
+    /// descend.
+
+    fn descend_bi(&self, op: Arc<dyn Fn(To) -> To>) -> Self {
+        let (children, ctx) = self.biplate();
+        ctx(children.map(op))
+    }
+
+    // TODO (niklasdewally): Uniplate does some wierd Haskell I don't understand here.
+    // What are children and universe are meant to be?
+
+    fn universe_bi(&self) -> im::Vector<To> {
+        self.children_bi()
+            .into_iter()
+            .flat_map(|child| child.universe())
+            .collect()
+    }
+
+    /// Returns the children of a type. If to == from then it returns the original element (in contrast to children).
+    fn children_bi(&self) -> im::Vector<To> {
+        self.biplate().0.list().0
+    }
+
+    fn transform_bi(&self, op: Arc<dyn Fn(To) -> To>) -> Self {
+        let (children, ctx) = self.biplate();
+        ctx(children.map(Arc::new(move |child| child.transform(op.clone()))))
+    }
+}
+
+pub trait Uniplate
+where
+    Self: Sized + Clone + Eq + 'static,
+{
+    fn uniplate(&self) -> (Tree<Self>, Box<dyn Fn(Tree<Self>) -> Self>);
+
+    fn descend(&self, op: Arc<dyn Fn(Self) -> Self>) -> Self {
+        let (children, ctx) = self.uniplate();
+        ctx(children.map(op))
+    }
+
+    /// Gest all children of a node, including itself and all children.
+    fn universe(&self) -> im::Vector<Self> {
+        let mut results = vector![self.clone()];
+        for child in self.children() {
+            results.append(child.universe());
+        }
+        results
+    }
+
+    /// Gets the direct children (maximal substructures) of a node.
+    fn children(&self) -> im::Vector<Self> {
+        let (children, _) = self.uniplate();
+        children.list().0.clone()
+    }
+
+    /// Reconstructs the node with the given children.
+    ///
+    /// # Returns
+    ///
+    /// If there are a different number of children given as there were originally, None is
+    /// returned.
+    fn with_children(&self, children: im::Vector<Self>) -> Option<Self> {
+        // 1. Turn old tree into list.
+        // 2. Check lists are same size.
+        // 3. Use the reconstruction function given by old_children.list() to
+        //   create a tree with the same structure but the new lists' elements .
+
+        let (old_children, ctx) = self.uniplate();
+        let (old_children_lst, rebuild) = old_children.list();
+        if old_children_lst.len() != children.len() {
+            None
```

Comment:
This could even be something louder. It should never happen.

Reply:
a Result<>?

Reply:
I'm tempted to say `panic`, as I don't think there is any reasonable recovery for the whole program if this changes -- and in general (this is my opinion, Oz may differ), I feel academic software should be happier with `panic`s than industry software, as while an unexpected exit is annoying, and unexpectedily incorrect answer because your software tried to carry on is much worse :)